sstables/sstables.cc

```diff
@@ -2242,6 +2246,19 @@ future<> sstable::move_from_staging(delayed_commit_changes* delay_commit) {
     return _storage.activate(*this, generation(), delay_commit);
 }
 
+future<> sstable::move_from_upload(generation_type new_generation, to_staging staging) {
```

Comment:
why have a function for every source location and every destination location?
how about a generic move or relocate function that gets the destination location (and maybe the source location if there are source-dependent side effects that need to take place. this can also be implemented with a given functor passed by the caller)

Reply:
There is one already, but on the storage level. For sstable level, currently the state changes are:
* any -> quarantine
* upload -> base
* upload -> staging
* staging -> base

And not all any -> any are meaningful, e.g. what would quarantine -> upload mean? I don't know :)

Having a generic function that can do any -> any would require it to validate that the transition is "supported", but what's the point? Also, some transitions require generation argument, so there will also need to be some "checks" for the provided generation to be OK.

Reply:
Re-implemented it this way. Target directory selection becomes ... pretty hairy :)